### PR TITLE
Add `scrollToEndButtonBehavior` style options

### LINKED ISSUE #3925: Add "More messages" button
## Feature Request

In addition to our existing "New messages"  button, which show up when there are new unread messages. We should add a new "More messages" button, which will be show up whenever the scroll view is not at the bottom, regardless of number of unread messages.

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #3925.

### Breaking changes

-  A new style option `scrollToEndButtonBehavior` is introduced to control when the scroll to end button should show (formerly "new messages" button):
   -  `styleOptions.hideScrollToEndButton` has been deprecated. To hide the scroll to end button, set `styleOptions.scrollToEndButtonBehavior` to `false`;
   -  `styleOptions.newMessageButtonFontSize` has been renamed to `styleOptions.scrollToEndButtonFontSize` to better reflect its purpose.

### Added

-  Resolves [#3925](https://github.com/microsoft/BotFramework-WebChat/issues/3925). Added `scrollToEndButtonBehavior` to control when the scroll to end button should show, removed `hideScrollToEndButton`, and renamed `newMessagesButtonFontSize` to `scrollToEndButtonFontSize`, by [@compulim](https://github.com/compulim), in PR [#3926](https://github.com/microsoft/BotFramework-WebChat/issues/3926).
   - Values for `scrollToEndButtonBehavior`:
      -  If unset, will maintain same behavior as previous versions, same as `"unread"`;
      -  `"unread"` will show when there are any unread and offscreen messages (default);
      -  `"any"` will show when there are any offscreen messages;
      -  `false` will always hide the button.
   -  Added new [`scrollToEndButtonMiddleware`](https://github.com/microsoft/BotFramework-WebChat/blob/main/packages/api/src/types/scrollToEndButtonMiddleware.ts) to customize the appearance of the scroll to end button.

## Description

A new style option is added to control how the scroll to end button should show (formerly "New messages" button).

Depends on the option, the button will be named "New messages" or "More messages" button.

With this PR, web developers can start customizing the look and behavior of the scroll to end button. It includes when their scroll to end button should show up. For example, they can always show their button regardless whether the scroll view is at the bottom or not.

## Design

Despite we listed breaking changes (due to deprecations), we have migration code for all breaking changes and web developers do not need to change anything when upgrading Web Chat. There will be warnings to remind them to update their code.

### Typing of middleware

Starting from this PR, we will add typings for our middleware, starting from `scrollToEndButtonMiddleware`.

Currently, we have 3 types of middleware:

- Renderer middleware, such as activity middleware, activity status middleware, avatar middleware, scroll to end button middleware, toast middleware, etc.
   - The middleware is for rendering UI components, through decoration/replacement/removal
   - Before final rendering, caller of the middleware can know whether the middleware will render or skip
      - This is useful in scenario like skip rendering event activity and not emitting empty `<li>`
      - This is core to activity status middleware, which requires 2-pass rendering to group avatar and activity status correctly
- Legacy renderer middleware, such as attachment middleware
   - In legacy mode, we assume the renderer will always render something
   - We should slowly migrate it to the renderer middleware
- Headless middleware, such as card action middleware
   - These middleware are for headless operations, such as submitting a card action, etc.

### `styleOptions.scrollToEndButtonBehavior`

The option can be set to 3 different values:

- `"unread"` will show when there are any unread and offscreen messages (default)
   - The button will show "New messages"
- `"any"` will show when there are any offscreen messages
   - The button will show "More messages"
- `false` will always hide the button

If unset, it will default to `"unread"`, which is the behavior of <= 4.13.0.

### `scrollToEndButtonMiddleware`

Sample code snippet to customize the appearance of the button. The code below will always show the scroll to end button regardless of scroll position (`atEnd` or not).

Please refer to [`scrollToEndButtonMiddleware.ts`](https://github.com/microsoft/BotFramework-WebChat/blob/main/packages/api/src/types/scrollToEndButtonMiddleware.ts) for the full signature of the middleware.

```js
const MyScrollToEndButton = ({ atEnd, onClick }) => {
  return (
    <button disabled={atEnd} onClick={onClick} type="button">
      {atEnd ? 'At bottom' : 'Not at bottom'}
    </button>
  );
};

renderWebChat(
  {
    directLine,
    scrollToEndButtonMiddleware: () => () => ({ atEnd }) => props => (
      <MyScrollToEndButton atEnd={atEnd} {...props} />
    ),
  },
  document.getElementById('app')
);
```

## Specific Changes

- Any references to "new messages" button will be renamed to "scroll to end" button
- Also added missing tests for localization override strings
   - This is to make sure the text inside the scroll to end button can be overrode

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] I have updated documentation

## Review Checklist

> This section is for contributors to review your work.

-  [x] ~Accessibility reviewed (tab order, content readability, alt text, color contrast)~
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] ~CSS styles reviewed (minimal rules, no `z-index`)~
-  [x] ~Documents reviewed (docs, samples, live demo)~
-  [x] Internationalization reviewed (strings, unit formatting)
-  [x] `package.json` and `package-lock.json` reviewed
-  [x] ~Security reviewed (no data URIs, check for nonce leak)~
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,11 +22,24 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Breaking changes
```
Comment (corinagum) [inline CHANGELOG.md:25]: 👍

Comment (corinagum) [reply CHANGELOG.md:25]: File a to do / add comments for when deprecation will be removed? (See styleOptions deprecation warnings)

File: packages/test/page-object/src/globals/pageObjects/scrollToTop.js [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+ */
 export default async function scrollToTop(offset = 0) {
   // Before scrolling to top, wait until the scroll is stabilized.
   await scrollStabilized();
 
   const transcriptScrollable = getTranscriptScrollableElement();
+  const maxOffset = transcriptScrollable.scrollHeight - transcriptScrollable.offsetHeight;
 
-  offset = offset < 0 ? transcriptScrollable.scrollHeight - transcriptScrollable.offsetHeight + offset : offset;
+  // Handles any numbers, 0, -0, Infinity, and -Infinity.
```
Comment (corinagum) [inline packages/test/page-object/src/globals/pageObjects/scrollToTop.js:21]: 👍 for more comments :)

File: packages/test/page-object/src/globals/pageElements/focusedActivity.js [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+  }
+
+  return false;
+}
+
 export default function focusedActivity() {
-  return document.getElementById(transcript().getAttribute('aria-activedescendant'));
+  const activity = document.getElementById(transcript().getAttribute('aria-activedescendant'));
+
+  // Could be focusing on Adaptive Card or <input> inside the activity, we will consider it is focused.
```
Comment (corinagum) [inline packages/test/page-object/src/globals/pageElements/focusedActivity.js:18]:
```suggestion
  // If focus is on the Adaptive Card or <input> inside the activity, it is considered as focused.
```

File: packages/component/src/Middleware/ScrollToEndButton/createScrollToEndButtonMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+export default function createScrollToEndButtonMiddleware(): ScrollToEndButtonMiddleware[] {
+  return [
+    () => () => ({ atEnd, styleOptions: { scrollToEndButtonBehavior }, unread }) =>
+      !scrollToEndButtonBehavior
+        ? // Don't show the button when it is set to false.
+          false
+        : scrollToEndButtonBehavior === 'any'
+        ? // Show when the scroll view is not at the end, regardless of number of unread activities.
+          !atEnd && ScrollToEndButton
+        : // Show when the scroll view is not at the end, and there are some unread activities.
```
Comment (corinagum) [inline packages/component/src/Middleware/ScrollToEndButton/createScrollToEndButtonMiddleware.ts:14]:
```suggestion
        : // Show when the scroll view is not at the end of the transcript, and there are new/unread activities.
```

File: packages/api/src/utils/warnOnce.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+export default function warnOnce(message) {
```
Comment (corinagum) [inline packages/api/src/utils/warnOnce.js:1]: 👍

File: packages/api/src/defaultStyleOptions.ts [hunk 6]
```diff
@@ -84,7 +84,7 @@ const DEFAULT_OPTIONS: Required<StyleOptions> = {
   markdownRespectCRLF: true,
 
   // Scroll behavior
-  hideScrollToEndButton: false,
+  hideScrollToEndButton: undefined, // Deprecated as of 4.14.0. Use "scrollToEndButtonBehavior" instead.
```
Comment (corinagum) [inline packages/api/src/defaultStyleOptions.ts:87]: Add 'remove by' date

File: __tests__/setup/elements/getScrollToBottomButton.js [hunk 7]
```diff
@@ -5,5 +5,5 @@ import getTranscript from './getTranscript';
 export default async function getScrollToBottomButton(driver) {
   const transcript = await getTranscript(driver);
 
-  return await transcript.findElement(By.css('.webchat__scrollToEndButton'));
+  return await transcript.findElement(By.css('.webchat__scroll-to-end-button'));
```
Comment (corinagum) [inline __tests__/setup/elements/getScrollToBottomButton.js:8]: This is ok for BEM naming convention?

File: __tests__/html/scrollToEndButton.persistWhileCallingUseScrollTo.js [hunk 8]
```diff
@@ -0,0 +1,4 @@
+/** @jest-environment ./packages/test/harness/src/host/jest/WebDriverEnvironment.js */
+
+test('scroll to end button should not persist while calling useScrollTo', () =>
```
Comment (corinagum) [inline __tests__/html/scrollToEndButton.persistWhileCallingUseScrollTo.js:3]:
```suggestion
test('scroll to end button should persist while calling useScrollTo', () =>
```
?

File: __tests__/html/scrollToEndButton.tabOrder.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+/** @jest-environment ./packages/test/harness/src/host/jest/WebDriverEnvironment.js */
+
+describe('scroll to end button', () => {
+  test('tab order', () => runHTML('scrollToEndButton.tabOrder.html'));
```
Comment (corinagum) [inline __tests__/html/scrollToEndButton.tabOrder.js:4]:
```suggestion
  test('have correct tab order', () => runHTML('scrollToEndButton.tabOrder.html'));
```

File: __tests__/html/styleOptions.deprecated.newMessageButtonFontSize.html [hunk 10]
```diff
@@ -0,0 +1,52 @@
+<!DOCTYPE html>
```
Comment (corinagum) [inline __tests__/html/styleOptions.deprecated.newMessageButtonFontSize.html:1]: Do you want to check for warning messages for these kinds of tests?

Comment (corinagum) [reply __tests__/html/styleOptions.deprecated.newMessageButtonFontSize.html:1]: Leave comments / reference on when these tests should be deleted?

File: packages/api/src/normalizeStyleOptions.ts [hunk 11]
```diff
@@ -1,12 +1,25 @@
 import defaultStyleOptions from './defaultStyleOptions';
 import StyleOptions, { StrictStyleOptions } from './StyleOptions';
+import warnOnce from './utils/warnOnce';
+
+const hideScrollToEndButtonDeprecation = warnOnce(
+  '"styleOptions.hideScrollToEndButton" has been deprecated. To hide scroll to end button, set "scrollToEndBehavior" to false. This deprecation migration will be removed on or after 2021-06-02.'
+);
+
+const newMessagesButtonFontSizeDeprecation = warnOnce(
+  '"styleOptions.newMessagesButtonFontSize" has been renamed to "styleOptions.scrollToEndButtonFontSize". This deprecation migration will be removed on or after 2021-06-02.'
```
Comment (corinagum) [inline packages/api/src/normalizeStyleOptions.ts:10]: 2023, right? or 2022?

**Review by corinagum — COMMENTED** [review]
Few comments / questions. Do you have time for 1:1?

**Review by corinagum — APPROVED** [review]
Approved pending comments/questions :)

File: packages/api/src/normalizeStyleOptions.ts (hunk 11 above)
Comment (compulim) [reply packages/api/src/normalizeStyleOptions.ts:10]: 2023. 😅

File: packages/api/src/utils/warnOnce.js (hunk 5 above)
Comment (compulim) [reply packages/api/src/utils/warnOnce.js:1]: Should move all deprecation warnings to this model. 😉

File: __tests__/html/styleOptions.deprecated.newMessageButtonFontSize.html (hunk 10 above)
Comment (compulim) [reply __tests__/html/styleOptions.deprecated.newMessageButtonFontSize.html:1]: Will leave comment on date.

There is an expectation at the end of the file to say we must show a warning, `{ expectDeprecations: true }`.

File: __tests__/setup/elements/getScrollToBottomButton.js (hunk 7 above)
Comment (compulim) [reply __tests__/setup/elements/getScrollToBottomButton.js:8]: Yap, this is BEM naming and this is B-only.

File: __tests__/html/scrollToEndButton.persistWhileCallingUseScrollTo.js (hunk 8 above)
Comment (compulim) [reply __tests__/html/scrollToEndButton.persistWhileCallingUseScrollTo.js:3]: 😅 You are right.

**Pull request merged**